src/hist/stats.py

```diff
@@ -0,0 +1,330 @@
+            raise ValueError(
+                "Cannot compute KS from histograms with different binning, use equal_bins=False"
+            )
+        if not np.allclose(self.axes[0].edges, other.axes[0].edges):
+            raise ValueError(
+                "Cannot compute KS from histograms with different binning, use equal_bins=False"
+            )
+
+    if mode not in ["auto", "exact", "asymp"]:
+        raise ValueError(f"Invalid value for mode: {mode}")
+    alternative = {"t": "two-sided", "g": "greater", "l": "less"}.get(
+        alternative.lower()[0], alternative
+    )
+    if alternative not in ["two-sided", "less", "greater"]:
+        raise ValueError(f"Invalid value for alternative: {alternative}")
+
+    data1 = self.values(flow=True)
+    data2 = other.values(flow=True)
+    n1 = data1.sum(dtype=int)
+    n2 = data2.sum(dtype=int)
+    edges1 = self.axes[0].centers
+    edges2 = other.axes[0].centers
+    cdf1 = np.cumsum(data1) / n1
+    cdf2 = np.cumsum(data2) / n2
+    if flow:
+        edges1 = np.concatenate([[-np.inf], edges1, [np.inf]])
+        edges2 = np.concatenate([[-np.inf], edges2, [np.inf]])
+
+    if equal_bins:
+        cdflocs = edges1
+
+        if not flow:
+            cdf1 = cdf1[1:-1]
+            cdf2 = cdf2[1:-1]
+
+        cddiffs = cdf1 - cdf2
+
+    elif not equal_bins:
+        edges_all = np.sort(np.concatenate([edges1, edges2]))
+        cdflocs = edges_all
+
+        if flow:
+            cdf1 = np.concatenate([[0], cdf1])
+            cdf2 = np.concatenate([[0], cdf2])
+
+        # Use np.searchsorted to get the CDF values at all bin edges
+        cdf1_all = cdf1[np.searchsorted(edges1, edges_all, side="right")]
+        cdf2_all = cdf2[np.searchsorted(edges2, edges_all, side="right")]
+        cddiffs = cdf1_all - cdf2_all
+
+    # Identify the location of the statistic
+    argminS = np.argmin(cddiffs)
+    argmaxS = np.argmax(cddiffs)
+    loc_minS = cdflocs[argminS]
+    loc_maxS = cdflocs[argmaxS]
+
+    # Ensure sign of minS is not negative.
+    minS = np.clip(-cddiffs[argminS], 0, 1)
+    maxS = cddiffs[argmaxS]
+
+    if alternative == "less" or (alternative == "two-sided" and minS > maxS):
+        d = minS
+        d_location = loc_minS
+        d_sign = -1
+    else:
+        d = maxS
+        d_location = loc_maxS
+        d_sign = 1
+
+    g = np.gcd(int(n1), int(n2))
+    n1g = n1 // g
+    n2g = n2 // g
+    pvalue = -np.inf
+
+    if mode == "auto":  # Always select exact
+        mode = "exact"
+    if mode == "exact" and n1g >= np.iinfo(np.int32).max / n2g:
+        # If lcm(n1, n2) is too big, switch from exact to asymp
+        mode = "asymp"
+        warnings.warn(
+            f"Exact ks_2samp calculation not possible with samples sizes "
+            f"{n1} and {n2}. Switching to 'asymp'.",
+            RuntimeWarning,
+            stacklevel=3,
+        )
+
+    if mode == "exact":
+        success, d, pvalue = _attempt_exact_2kssamp(n1, n2, g, d, alternative)
```

Comment:
FYI, this really worries me. Using a hidden object is a really bad idea - see https://github.com/copier-org/copier/issues/1225 for the most recent time I've seen someone broken by this! Is it possible to avoid this?

Reply:
If you mean the function written by scipy, I was trying to avoid writing too many lines since after defining the test statistic from the histograms, performing the test is the same. I could write a similar function and have it here in hist but it would be mainly done by copying code from scipy

Reply:
I've just added the code in the `stats.py` file instead of importing from `scipy`. Do you think this is resolved?